flashinfer/artifacts.py

```diff
@@ -0,0 +1,100 @@
+"""
+Copyright (c) 2025 by FlashInfer team.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+  http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+"""
+
+import os
+import re
+import time
+from concurrent.futures import ThreadPoolExecutor, as_completed
+
+import requests
+
+from .jit.core import logger
+from .jit.cubin_loader import FLASHINFER_CUBINS_REPOSITORY, get_cubin
+
+
+def get_available_cubin_files(source, retries=3, delay=5, timeout=10):
+    for attempt in range(1, retries + 1):
+        try:
+            response = requests.get(source, timeout=timeout)
+            response.raise_for_status()
+            hrefs = re.findall(r'\<a href=".*\.cubin">', response.text)
+            files = [(h[9:-8], ".cubin") for h in hrefs]
+            return files
+
+        except requests.exceptions.RequestException as e:
+            logger.warning(
+                f"Fetching available files {source}: attempt {attempt} failed: {e}"
+            )
+
+            if attempt < retries:
+                logger.info(f"Retrying in {delay} seconds...")
+                time.sleep(delay)
+            else:
+                logger.error("Max retries reached. Fetch failed.")
+                return []
+
+
+class ArtifactPath:
+    TRTLLM_GEN_FMHA: str = "52e676342c67a3772e06f10b84600044c0c22b76/fmha/trtllm-gen/"
+    TRTLLM_GEN_BMM: str = (
+        "991e7438224199de85ef08a2730ce18c12b4e0aa/batched_gemm-c603ed2-2dc78d9/"
+    )
+    TRTLLM_GEN_GEMM: str = (
+        "fffd607babb0844f24225997409747ca38229333/gemm-c603ed2-f2b0c24/"
+    )
+    CUDNN_SDPA: str = "4c623163877c8fef5751c9c7a59940cd2baae02e/fmha/cudnn/"
+    DEEPGEMM: str = "d25901733420c7cddc1adf799b0d4639ed1e162f/deep-gemm/"
+
+
+class MetaInfoHash:
+    TRTLLM_GEN_FMHA: str = (
+        "8c5630020c0452fb1cd1ea7e3b8fdbb7bf94f71bd899ed5b704a490bdb4f7368"
+    )
+    DEEPGEMM: str = "69aa277b7f3663ed929e73f9c57301792b8c594dac15a465b44a5d151b6a1d50"
+
+
+def download_artifacts() -> bool:
+    env_backup = os.environ.get("FLASHINFER_CUBIN_CHECKSUM_DISABLED", None)
+    os.environ["FLASHINFER_CUBIN_CHECKSUM_DISABLED"] = "1"
```

Comment:
Can you document why you're doing this kind of manipulations?

Reply:
There is a risk here that there is some issue in the download and we produce a container which wouldn't have the right cubins and need to redownload later.